amaru/src/mempool/mod.rs

```diff
@@ -0,0 +1,104 @@
+use crate::ledger;
+use pallas_primitives::conway::Tx;
+use std::collections::BTreeSet;
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+// State
+// ----------------------------------------------------------------------------
+
+/// The state of the mempool is split into two sub-components:
+///
+/// - A _ledger_state_ which contains a view of a ledger state, which is used to validate
+///   transactions.
+///
+/// - An _ephemeral_ state, which is maintained as a sequence of diff operations to be applied on
+///   top of the _ledger_state_. It should be re-calculated as transactions are added or removed
+///   from the mempool.
+pub struct State<S, E>
+where
+    S: ledger::store::Store<Error = E>,
+{
+    /// The current view of the ledger state.
+    ledger_state: ledger::state::State<S, E>,
+
+    /// The ephemeral state, which is applied on top of the ledger_state using transactions
+    /// in the mempool.
+    ephemeral: ledger::state::VolatileDB,
+}
+
+/// A transaction in the mempool.
+pub trait MempoolTx<M>: Ord + PartialEq {
+    /// Any additional _metadata_ associated with the transaction. For example, it could include a timestamp
+    /// of when it was received, a reference to the peer that sent it, or some priority value used
+    /// to order transactions.
+    fn metadata(&self) -> Option<M> {
+        None
+    }
+
+    /// The transaction itself.
+    fn tx(&self) -> &Tx;
+}
+
+/// The mempool is a queue of transactions that are waiting to be validated and applied to an
+/// ephemeral ledger state.
+pub struct Mempool<T, M>
+where
+    T: MempoolTx<M>,
+{
+    /// The queue of transactions. The ordering of transactions are defined by the Ord trait of the
+    /// type T and left to the implementer.
+    transactions: Arc<RwLock<BTreeSet<T>>>,
+
+    _marker: std::marker::PhantomData<M>,
+}
+
+impl<T: MempoolTx<M>, M> Default for Mempool<T, M> {
+    fn default() -> Self {
+        Mempool {
+            transactions: Arc::new(RwLock::new(BTreeSet::new())),
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<T: MempoolTx<M>, M> Mempool<T, M> {
```

Comment:
I think what would help me (and others) better review this design is to provide some kind of model embodying the expected properties and behaviour of the mempool as tests. This might strengthen the case for the data structures used or on the contrary suggests we might use something else, and would expose more clearly the interaction between mempool and other parts of the system.
For example, how does removing a transaction plays with the ledger state? When a tx is added/removed, what happens to the `State`? In general, what are the invariant(s) we want the mempool to guarantee b/w State and the queue of transactions it holds? 
This does not require real transactions and does not need to be super involved, but I think it would help to expose potential shortcomings of the design, or validate it.

Reply:
I can move forward with it and start implementing, but what I wanted to avoid by making this a `Draft` PR was doing a ton of work and then people criticizing the design late in the game. I wanted to get early feedback on it to make sure it's on the right track.

I agree that it would make more sense if there were an example implementation where you could see what happens on block rollbacks for example.  This will be the next phase for this PR, but I wanted to make sure I'm at least somewhat in the ballpark before moving forward.

Reply:
> doing a ton of work and then people criticizing the design late in the game

I can relate. I would not go for a full blown implementation, but having a small model to play with and understanding the dynamics of the proposed design would help, ie. a couple of tests showing how the interfaces are used, how the state evolves, what invariants we need to maintain...